portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -77,24 +88,61 @@ private String toJSONString(Map<String, Map<String, String>> map, String name) {
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get the input from the form.
     String feedback = request.getParameter("text-input");
+    // Get the URL of the image that the user uploaded to Blobstore.
+    String imageUrl = getUploadedFileUrl(request, "image");
     long timestamp = System.currentTimeMillis();
     String userEmail = userService.getCurrentUser().getEmail();
 
     // Only add feedback if valid input.
-    if (feedback == "") {
-      LOGGER.warning("No input");
-      return;
-    } else {
+    if (feedback != "" || imageUrl != null) {
       Entity commentEntity = new Entity("Comment");
 
       commentEntity.setProperty("comment", feedback);
       commentEntity.setProperty("timestamp", timestamp);
       commentEntity.setProperty("userEmail", userEmail);
 
+      if (imageUrl != null) {
+        commentEntity.setProperty("imageUrl", imageUrl);
+      } else {
+        commentEntity.setProperty("imageUrl", "null");
+      }
       datastore.put(commentEntity);
     }
 
     // Redirect back to the HTML page.
     response.sendRedirect("/index.html");
   }
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // Use the relative path to the image.
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
```

Comment:
Please add a comment explaining why we need to apply this manipulation.

Reply:
Added a comment